rotkehlchen/externalapis/cowswap.py

```diff
@@ -0,0 +1,94 @@
+import json
+import logging
+from typing import TYPE_CHECKING, Any, Final
+
+import requests
+
+from rotkehlchen.db.settings import CachedSettings
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import SupportedBlockchain, Timestamp
+from rotkehlchen.utils.misc import iso8601ts_to_timestamp
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+ORDER_TYPE_CHANGE_TS: Final = Timestamp(1710806400)
+CHAIN_MAPPING: Final = {
+    SupportedBlockchain.ETHEREUM: 'mainnet',
+    SupportedBlockchain.GNOSIS: 'xdai',
+    SupportedBlockchain.ARBITRUM_ONE: 'arbitrum_one',
+}
+
+
+class CowswapAPI:
+    """This is the Cowswap API interface.
+
+    https://api.cow.fi/docs/
+    """
+
+    def __init__(self, database: 'DBHandler', chain: 'SupportedBlockchain') -> None:
+        self.database = database
+        self.session = requests.session()
+
+        if chain not in CHAIN_MAPPING:
+            log.warning(f'Encountered unsupported Cowswap API chain {chain}. Falling back to mainnet.')  # noqa: E501
+            chain = SupportedBlockchain.ETHEREUM
+
+        self.api_url = f'https://api.cow.fi/{CHAIN_MAPPING[chain]}/api/v1'
+
+    def _query(self, endpoint: str) -> dict[str, Any]:
+        """Query cowswap API endpoint."""
+        query_str = f'{self.api_url}/{endpoint}'
+        log.debug(f'Querying Cowswap API {query_str}')
+        timeout = CachedSettings().get_timeout_tuple()
+        try:
+            response = self.session.get(url=query_str, timeout=timeout)
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'Querying {query_str} failed due to {e!s}') from e
+
+        if response.status_code != 200:
+            raise RemoteError(f'Cowswap API request {response.url} failed with HTTP status code {response.status_code} and text {response.text}')  # noqa: E501
+
+        try:
+            json_ret = json.loads(response.text)
+        except json.decoder.JSONDecodeError as e:
+            raise RemoteError(f'Cowswap API returned invalid JSON response: {response.text}') from e  # noqa: E501
+
+        return json_ret
+
+    def query_order_data(self, order_uid: str) -> tuple[str, str]:
+        """Query offchain order data from the database or the api.
+        Returns raw_fee_amount and order_type as a tuple.
+        """
+        with self.database.conn.read_ctx() as cursor:
+            result = cursor.execute(
+                'SELECT order_type, raw_fee_amount FROM cowswap_orders WHERE identifier = ?',
+                (order_uid,),
+            ).fetchone()
+
+        if result is not None:
+            raw_fee_amount = result[0]
+            order_type = result[1]
+        else:
+            data = self._query(f'orders/{order_uid}')
```

Comment:
In such cases it's much easier to exit early **and** not have indented code.


```suggestion
        if result is not None:
            return int(result[0]), result[1]

        # else, we need to query the API
        data = self._query(f'orders/{order_uid}')
        ....
```